modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/resources/META-INF/resources/js/components/PageRenderer/index.js

```diff
@@ -0,0 +1,249 @@
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import {ClayIconSpriteContext} from '@clayui/icon';
+import core from 'metal';
+import Soy from 'metal-soy';
+import React from 'react';
+
+import {PageProvider} from '../../hooks/usePage.es';
+import {getConnectedReactComponentAdapter} from '../../util/ReactComponentAdapter.es';
+import * as DefaultVariant from './DefaultVariant.es';
+import * as EditablePageHeader from './EditablePageHeader.es';
+import {Layout} from './Layout.es';
+import * as MultiPages from './MultiPagesVariant.es';
+import * as Paginated from './PaginatedVariant.es';
+import * as SuccessPage from './SuccessVariant.es';
+import * as Tabbed from './TabbedVariant.es';
+import * as Wizard from './WizardVariant.es';
+import templates from './index.soy';
+
+const LAYOUT_TYPES = {
+	MULTI_PAGES: 'multi_pages',
+	PAGINATED: 'paginated',
+	SINGLE_PAGE: 'single-page',
+	SUCCESS: 'success',
+	TABBED: 'tabbed',
+	WIZARD: 'wizard',
+};
+
+const LAYOUT_COMPONENTS_TYPES = {
+	[LAYOUT_TYPES.MULTI_PAGES]: MultiPages,
+	[LAYOUT_TYPES.PAGINATED]: Paginated,
+	[LAYOUT_TYPES.SINGLE_PAGE]: {
+		Container: DefaultVariant.Container,
+	},
+	[LAYOUT_TYPES.SUCCESS]: SuccessPage,
+	[LAYOUT_TYPES.TABBED]: Tabbed,
+	[LAYOUT_TYPES.WIZARD]: Wizard,
+};
+
+const PAGE_HEADER_TYPES = {
+	EDITABLE: 'editable',
+};
+
+const PAGE_HEADER_COMPONENT_TYPE = {
+	[PAGE_HEADER_TYPES.EDITABLE]: EditablePageHeader.PageHeader,
+};
+
+const isEmptyPage = ({rows}) => {
+	let empty = false;
+
+	if (!rows || !rows.length) {
+		empty = true;
+	}
+	else {
+		empty = !rows.some(({columns}) => {
+			let hasFields = true;
+
+			if (!columns) {
+				hasFields = false;
+			}
+			else {
+				hasFields = columns.some((column) => column.fields.length);
+			}
+
+			return hasFields;
+		});
+	}
+
+	return empty;
+};
+
+const normalizePage = (page, editingLanguageId) => {
+	if (core.isObject(page.description)) {
+		page = {
+			...page,
+			description: page.description[editingLanguageId],
+		};
+	}
+	if (core.isObject(page.title)) {
+		page = {
+			...page,
+			title: page.title[editingLanguageId],
+		};
+	}
+
+	return page;
+};
+
+const getVariant = ({page, pages, paginationMode, view, viewMode}) => {
+	if (
+		page.contentRenderer &&
+		page.contentRenderer !== null &&
+		page.contentRenderer !== ''
+	) {
+		return page.contentRenderer;
+	}
+	else if (paginationMode === null || paginationMode === '') {
+		if (pages.length > 1) {
+			return LAYOUT_TYPES.WIZARD;
+		}
+		else {
+			return LAYOUT_TYPES.SINGLE_PAGE;
+		}
+	}
+	else if (
+		paginationMode === LAYOUT_TYPES.WIZARD &&
+		!viewMode &&
+		view === 'formBuilder'
+	) {
+		return LAYOUT_TYPES.MULTI_PAGES;
+	}
```

Comment:
This is one of the strange things I had to bring from Soy, apparently, we are doing a replacement of the `paginationMode` hardcoded on the frontend. The value comes with `wizard` and we ignore it here. I think this should not be done on the frontend but from the backend, we are practically not respecting `paginationMode`.

Reply:
Agree. We should talk to the Forms team about it.